src/ejabberd_c2s_info_handler.erl

```diff
@@ -0,0 +1,56 @@
+-module(ejabberd_c2s_info_handler).
+
+-include("ejabberd_c2s.hrl").
+-include("mongoose.hrl").
+
+% API for other modules
+-export([add/5, remove/2]).
+
+% API for ejabberd_c2s
+-export([get_for/2,
+         add_to_state/5,
+         remove_from_state/2,
+         safe_call/3,
+         safe_call/6]).
+
+add(C2S, Tag, M, F, HandlerState) ->
+    p1_fsm_old:send_all_state_event(C2S, {add_info_handler, Tag, M, F, HandlerState}).
+
+remove(C2S, Tag) ->
+    p1_fsm_old:send_all_state_event(C2S, {remove_info_handler, Tag}).
+
+get_for(Tag, #state{ info_handlers = InfoHandlers }) ->
+    maps:get(Tag, InfoHandlers, undefined).
+
+add_to_state(Tag, M, F, HandlerState, #state{ info_handlers = InfoHandlers0 } = C2SState) ->
+    C2SState#state{ info_handlers = InfoHandlers0#{ Tag => {M, F, HandlerState} } }.
+
+remove_from_state(Tag, #state{ info_handlers = InfoHandlers0 } = C2SState) ->
+    C2SState#state{ info_handlers = maps:remove(Tag, InfoHandlers0) }.
+
+safe_call(Tag, Data, C2SState) ->
+    case get_for(Tag, C2SState) of
+        {M, F, HandlerState} ->
+            case safe_call(Tag, Data, M, F, HandlerState, C2SState) of
+                {error, _} ->
+                    C2SState;
+                NC2SState0 ->
+                    NC2SState0
+            end;
+        _ ->
+            undefined
+    end.
+
+safe_call(Tag, Data, M, F, HandlerState, C2SState) ->
+    try M:F(Data, HandlerState, C2SState) of
+        {NewHandlerState, NewC2SState0} ->
```

Comment:
I somehow don't like allowing other modules to mess with internal state of c2s process. They have to be able to read it, that's for sure, but if they need to modify it they have their own state for that. For example, in a while I'll be refactoring mod_roster to use it, and my gut feeling is that if it is mod_roster that is responsible for handling roster, then all roster-related functions should go there (like current `ejabberd_c2s:roster_change/4`), and the whole roster should be stored in the handler state, made available via a hook handler.

Reply:
If handler state is available for this handler only, then it makes extensibility harder.
It's good idea to pass the whole c2s state and have getters and setters to get and set specific handler state.
So we would still have incapsulation when we need and we can violate the encapsulation when we _really_ need it (to access state of one handler from another, for example).

Reply:
The primary thought behind this code was to allow custom handlers to read & write "aux fields" from C2S state but it may indeed be replaced by new aux fields API and passing only them to the handler instead of C2S state (or like you've mentioned: pass it as read-only).

Reply:
I think that passing only an auxiliary field of the c2s state would be fine from the encapsulation perspective, and if we need to modify this we can just rebind it to the c2s_state outside of the handlers. If a handler needs to rebind other fields on the c2s state... Dunno, I'm having in mind something similar to what we do with `mongoose_acc`, like,
```erlang
    try M:F(Data, HandlerState) of
        NewHandlerState ->
              OtherStuff = extract_important_variables_from_handlers(NewHandlerState),
              NewC2SState = merge_states(C2SState, OtherStuff);
        ...
```

Reply:
To rephrase - IMO:
* a handler MUST be able to read c2s process state
* a handler MUST NOT be able to modify c2s process state
* a handler MUST NOT be able to modify other handlers' state
now, there is an alternative:
EITHER
* a handler can read state of other handlers
* the c2s process can read state of all handlers
OR
* handler state is available to this handler only, but the handler MAY expose its state (or part of it) by creating a hook handler

Reply:
I think we're better off keeping encapsulation stricter for now. Maybe we can make handlers only use `HandlerState` by default, but opt into getting the complete `C2SState` for some specific handlers that require it? But keeping "explicit is better than implicit" in mind.

Reply:
I'm in favor of passing only the `HandlerState` to the handler. If a handler needs access to the stat of the C2S process it can be added with an explicit demand for it, like @erszcz suggested.
I also agree with the statement that a handler is not allowed to modify the C2S state, only it's own state.